tox.ini

```diff
@@ -13,17 +13,43 @@ deps =
     pytest-subtests
 extras =
     {env:TEST_EXTRAS_TOX:}
-setenv = COVERAGE_FILE={toxworkdir}/.coverage.{envname}
+setenv =
+    COVERAGE_FILE={toxworkdir}/.coverage/.coverage.{envname}
 commands =
     pytest {toxinidir} {posargs: --cov=zoneinfo --cov=tests}
 
+# This should probably be integrated into the main testenv as an option if such
+# a thing is possible, but because it's more important that it runs on CI and
+# it's not necessarily very cross-platform, I've split it out for now.
+[testenv:gcov]
+description = Run the tests and collect C coverage stats
+deps =
+    gcovr
+    hypothesis>=5.7.0
+    pytest
+    pytest-subtests
+extras =
+    tzdata
+passenv =
+    CC
+setenv =
+    CFLAGS=--coverage -fprofile-dir={envdir}/coverage
+    LDFLAGS=--coverage
+    GCNO_TARGET_DIR={envdir}/coverage
+commands =
+    pytest {toxinidir}
+    gcovr -r {toxinidir} -k
+    python -c "import os; os.makedirs('{toxworkdir}/.coverage', exist_ok=True)"
+    gcovr -r {toxinidir} --xml \
+          -o {toxworkdir}/.coverage/.gcov_coverage.{envname}.xml
+
 [testenv:coverage-report]
 basepython = python3.8
 skip_install = true
 deps = coverage[toml]>=5.0.2
 depends = py38
 setenv=COVERAGE_FILE=.coverage
-changedir = {toxworkdir}
+changedir = {toxworkdir}/.coverage
```

Comment:
Same comment about upgrading the tests (==> do it separately) vs C-specific (==> did you break the python version) as above, and repeat for further changes in this file.